src/Connection/DefaultConnectionPool.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+namespace Amp\Http\Client\Connection;
+
+use Amp\ByteStream\StreamException;
+use Amp\CancellationToken;
+use Amp\CancelledException;
+use Amp\Http\Client\Internal\CombinedCancellationToken;
+use Amp\Http\Client\Request;
+use Amp\Http\Client\SocketException;
+use Amp\Http\Client\TimeoutException;
+use Amp\Promise;
+use Amp\Socket;
+use Amp\Socket\ClientTlsContext;
+use Amp\Socket\ConnectContext;
+use Amp\Socket\Connector;
+use Amp\Socket\EncryptableSocket;
+use Amp\TimeoutCancellationToken;
+use function Amp\call;
+
+final class DefaultConnectionPool implements ConnectionPool
+{
+    /** @var Connector */
+    private $connector;
+
+    /** @var Promise[][] */
+    private $connections = [];
+
+    public function __construct(?Connector $connector = null)
+    {
+        $this->connector = $connector ?? Socket\connector();
+    }
+
+    public function getConnection(Request $request, CancellationToken $cancellation): Promise
+    {
+        return call(function () use ($request, $cancellation) {
+            $uri = $request->getUri();
+            $isHttps = $uri->getScheme() === 'https';
+            $defaultPort = $isHttps ? 443 : 80;
+
+            $authority = $uri->getHost() . ':' . ($uri->getPort() ?: $defaultPort);
+            $key = $uri->getScheme() . '://' . $authority;
+
+            if (!empty($this->connections[$key])) {
+                foreach ($this->connections[$key] as $index => $connection) {
+                    $connection = yield $connection;
+                    \assert($connection instanceof Connection);
+
+                    if (!$connection->isBusy()) {
+                        return $connection;
+                    }
+                }
+
+                ++$index;
```

Comment:
This isn't safe. Suppose we have 3 concurrent connections which all allow just one request. Due to the `yielding` above, the second and third both have `$index = 0` and hang in the `yield` until the connection is established. Both push to `$index = 1` then.

Reply:
Ah, right, since the new connections won't be added to the list used by foreach. Hmm… might have to use a queue or something where that would be the case.

Reply:
I'd just do a keyless push and query the last key.